File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/api/ProducerRequest.java
package kafka.api;

import kafka.message.ByteBufferMessageSet;
import kafka.network.Request;
import kafka.utils.Utils;

import java.nio.ByteBuffer;
import java.util.function.Function;

public class ProducerRequest extends Request {
    public static int RandomPartition = -1;
    
    public String topic;
    public int partition;
    public  ByteBufferMessageSet messages;
    
    public ProducerRequest(String topic, int partition, ByteBufferMessageSet messages) {
        this.topic = topic;
        this.partition = partition;
        this.messages = messages;
    }
    
    
    public static ProducerRequest readFrom(ByteBuffer buffer) {
        String topic = Utils.readShortString(buffer, "UTF-8");
        int partition = buffer.getInt();
        int messageSetSize = buffer.getInt();
        ByteBuffer messageSetBuffer = buffer.slice();
        messageSetBuffer.limit(messageSetSize);
        buffer.position(buffer.position() + messageSetSize);
        try {
            return new ProducerRequest(topic, partition, new ByteBufferMessageSet(messageSetBuffer));
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
        return null;
    }
    
    public void writeTo(ByteBuffer buffer) {
        
        Utils.writeShortString(buffer, topic, "UTF-8");
        buffer.putInt(partition);
        buffer.putInt(messages.serialized().limit());
        buffer.put(messages.serialized());
        messages.serialized().rewind();
    }
    
    public int sizeInBytes() {
        return 2 + topic.length() + 4 + 4 + Integer.valueOf(messages.sizeInBytes() + "");
    }
    
    
    public int getTranslatedPartition(int randomSelector) {
        if (partition == ProducerRequest.RandomPartition) {
            return randomSelector;
        } else {
            return partition;
        }
    }
    
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("ProducerRequest(");
        builder.append(topic + ",");
        builder.append(partition + ",");
        builder.append(messages.sizeInBytes());
        builder.append(")");
        builder.toString();
        return builder.toString();
    }
    
    public boolean equals(Object other) {
        if (other instanceof ProducerRequest) {
            ProducerRequest that = (ProducerRequest) other;
            return (that.equals(this) && topic == that.topic && partition == that.partition && messages.equals(that.messages));
        } else {
            return false;
        }
    }
    
    public Boolean  canEqual(Object other ){
       return other instanceof  ProducerRequest ;
    }
    
    public   int hashCode() {
        return  31+(17*partition)+topic.hashCode() +messages.hashCode();
    }
    
}

File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/api/ProducerRequest.scala
/**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
  * this work for additional information regarding copyright ownership.
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
  * the License.  You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package kafka.api

import java.nio._
import kafka.message._
import kafka.network._
import kafka.utils._

object ProducerRequest {
  val RandomPartition = -1

  def readFrom(buffer: ByteBuffer): ProducerRequest = {
    val topic = Utils.readShortString(buffer, "UTF-8")
    val partition = buffer.getInt
    val messageSetSize = buffer.getInt
    val messageSetBuffer = buffer.slice()
    messageSetBuffer.limit(messageSetSize)
    buffer.position(buffer.position + messageSetSize)
    new ProducerRequest(topic, partition, new ByteBufferMessageSet(messageSetBuffer))
  }
}

class ProducerRequest(val topic: String,
                      val partition: Int,
                      val messages: ByteBufferMessageSet) extends Request(RequestKeys.Produce) {
    println("-------------ProducerRequest--------------------init 初始化ProducerRequest-------------------------")

  override def writeTo(buffer: ByteBuffer) {
    println("-------------ProducerRequest----------------------writeTo-------------buffer-对象-" + buffer)
    Utils.writeShortString(buffer, topic, "UTF-8") // 将topic的信息放入buffer
    buffer.putInt(partition)  // 将 partition的值放入 buffer
    buffer.putInt(messages.serialized.limit) // 将消息的大小放入buffer
    buffer.put(messages.serialized)
    println("-------------ProducerRequest----------------------writeTo-------------------------messages.serialized.limit-" +  messages.serialized.limit)
    println("-------------ProducerRequest----------------------writeTo-------------------------messages.serialized-" + new String(messages.serialized.array()))
    messages.serialized.rewind
  }

  override def sizeInBytes(): Int ={
    println("-------------ProducerRequest-------------------------sizeInBytes-------------messages.sizeInBytes =" + messages.sizeInBytes.asInstanceOf[Int])
    val dataSize = 2 + topic.length + 4 + 4 + messages.sizeInBytes.asInstanceOf[Int]
    println("-------------ProducerRequest-------------------------sizeInBytes--------- 2 + topic.length + 4 + 4=" + ( 2 + topic.length + 4 + 4))
    println("-------------ProducerRequest-------------------------sizeInBytes--------------------------dataSize=" + dataSize)
    dataSize
  }

  def getTranslatedPartition(randomSelector: String => Int): Int = {
    if (partition == ProducerRequest.RandomPartition)
      return randomSelector(topic)
    else
      return partition
  }

  override def toString: String = {
    val builder = new StringBuilder()
    builder.append("ProducerRequest(")
    builder.append(topic + ",")
    builder.append(partition + ",")
    builder.append(messages.sizeInBytes)
    builder.append(")")
    builder.toString
  }

  override def equals(other: Any): Boolean = {
    other match {
      case that: ProducerRequest =>
        (that canEqual this) && topic == that.topic && partition == that.partition &&
          messages.equals(that.messages)
      case _ => false
    }
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[ProducerRequest]

  override def hashCode: Int = 31 + (17 * partition) + topic.hashCode + messages.hashCode

}
